### DISCUSSION TITLE
Custom message routing

### DISCUSSION THREAD

**Discussion by jpbolduc85** [aws/aws-dotnet-messaging#164] [Q&A] [answered]
Hi guys, 

I've looked at this new AWS.Messaging it seems very promising :). I didn't know that such a project existed and we have been building our own solution over the last year  which is quite similar to this library but not as elaborated.  We have one important feature in our solution which is the message routing by messageSource, messageType and messageVersion. This allows us to route the same message type to differents handlers by combining all those properties. I was just wondering if there is a way with your solution to customize the routing base on different attributes like we do ? 

For ex:

`        
private IMessageHandler GetMessageHandler(IServiceProvider provider, string messageSource, string messageType, string messageVersion)
        {
            var handler = provider.GetServices<IMessageHandler>().Where(x => x.MessageSource == messageSource &&
                                                                             x.MessageType == messageType && 
                                                                             x.MessageVersion == messageVersion).SingleOrDefault();

            return handler ?? throw new InvalidOperationException($"No message handler found for message type: {messageType}, message source: {messageSource}, message version: {messageVersion}");
        }

    public abstract class MessageHandler<T> : IMessageHandler where T : IMessage, new()
    {
        public string MessageType { get; }
        public string MessageSource { get; }
        public string MessageVersion { get; }

        protected readonly ILogger _logger;

        public MessageHandler(ILogger logger, string messageSource, string messageType, string messageVersion)
        {
            _logger = logger;
            MessageSource = messageSource;
            MessageType = messageType;
            MessageVersion = messageVersion;
        }

        public async Task HandleSQSMessageAsync(Amazon.SQS.Model.Message sqsMsg)
        {
            var message = JsonSerializer.Deserialize(sqsMsg.Body, typeof(T));
            if (message is null)
            {
                _logger.LogError("Message is not valid. MessageId: {0}, Body: {1}. Message will not be processed.", sqsMsg.MessageId, sqsMsg.Body);
                throw new InvalidOperationException();
            }
            await HandleMessageAsync((T)message);
        }

        public abstract Task HandleMessageAsync(T message);
    }
`

Thanks!

**Comment by bhoradc** [comment, marked as answer]
Hi @jpbolduc85, 

Thank you for your question. `AWS.Messaging` library is designed with flexibility in mind. The key point here is that most classes in this library are backed by interfaces. This design choice enables you to implement your own versions of these components, including custom routing logic, and then inject them into the dependency injection container.

If you find that the suggested approach of creating custom implementations doesn't fully meet your requirement, we encourage you to submit a feature request for this functionality. 

Regards, 
Chaitanya
